### PR TITLE
Add hard line break that is unmodifed by 'grouped'

### PR DESCRIPTION
**Pull request by seanmcl**
Addresses #76.  There are still a couple failing tests, but I figured I'd publish
my work so far so @olafurpg can see progress is being made, and
others can take a look.

### CODE REVIEW

**Comment by seanmcl** [comment]
All tests passing now.  I'll come up with some more Scala examples for @olafurpg's use case.

File: core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala [hunk 1]
```diff
@@ -304,9 +302,5 @@ class PaigesScalacheckTest extends FunSuite {
       val expected = Doc.text(new String(Array(c1, c2)))
       assert(got === expected)
     }
-
-    // here is a hard case:
-    assert((Doc.char('a') + Doc.char('\n')) === Doc.text("a\n"))
```
Comment (seanmcl) [inline core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala]: This test has undefined behavior.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 2]
```diff
@@ -442,60 +444,39 @@ sealed abstract class Doc extends Product with Serializable {
...
-   * Convert this Doc to a single-line representation.
-   *
-   * All newlines are replaced with spaces (and optional indentation
-   * is ignored). The resulting Doc will never render any newlines, no
-   * matter what width is used.
-   */
-  def flatten: Doc = flattenOption getOrElse this
-
-  // return the flattened doc, and if it is different
-  private def flattenBoolean: (Doc, Boolean) = {
```
Comment (johnynek) [inline core/src/main/scala/org/typelevel/paiges/Doc.scala]: previously we were using that flattenBoolean simplifies the Doc also (it makes sure we have the correct associativity on Concat, and it removes needless Align, Nest, etc...

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 3]
```diff
@@ -442,60 +444,39 @@ sealed abstract class Doc extends Product with Serializable {
...
-
-    type DB = (Doc, Boolean)
+  // Try to put the document on a single line.  Will succeed iff there are no
+  // hard line breaks.
+  // Invariant: If flatten(d) == Some(d') then d' has no FlatAlt nodes
+  private[paiges] def flatten: Option[Doc] = {
+    sealed trait Flatten extends Product with Serializable
```
Comment (johnynek) [inline core/src/main/scala/org/typelevel/paiges/Doc.scala]: this change seems orthogonal to the rest. Can we pull out a small change that just introduces this `Flatten` type to clean the code up?

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 4]
```diff
@@ -442,60 +444,39 @@ sealed abstract class Doc extends Product with Serializable {
...
+        case Line => No
+        case Nest(_, d) => loop(Flat(d, h.changed), stack, front) // no Line, so Nest is irrelevant
+        case Align(d) => loop(Flat(d, h.changed), stack, front) // no Line, so Align is irrelevant
+        case Union(a, _) => loop(Flat(a, true), stack, front) // invariant: flatten(union(a, b)) == flatten(a)
+        case Concat(a, b) => loop(Flat(a, h.changed), Flat(b, h.changed) :: stack, front)
+        case FlatAlt(_, b) => loop(Flat(b, true), stack, front)
       }
-    loop((this, false), Nil, Nil)
+
+    loop(Flat(this, false), Nil, Nil) match {
```
Comment (johnynek) [inline core/src/main/scala/org/typelevel/paiges/Doc.scala]: I don't think we want to hide the result when nothing changes since you could have removed Align or Nest does, and you cleaned up Concats.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 5]
```diff
@@ -534,13 +516,18 @@ object Doc {
    */
   private[paiges] case object Empty extends Doc
 
+  /**
+    * Render 'default' except when flattened.
+    * Invariant: isEmpty(FlatAlt(a)) <=> isEmpty(FlatAlt(b))
```
Comment (johnynek) [inline core/src/main/scala/org/typelevel/paiges/Doc.scala]: do you mean `isEmpty(a) == isEmpty(b)`?

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 6]
```diff
@@ -721,6 +715,15 @@ object Doc {
   def split(str: String, pat: Regex = Doc.splitWhitespace, sep: Doc = Doc.lineOrSpace): Doc =
     foldDocs(pat.pattern.split(str, -1).map(Doc.text))((x, y) => x + (sep + y))
 
+  /**
+   * Render whenFlat if the document (or enclosing document) is grouped.
+   * Otherwise render default.
+   *
+   * Invariant: width(default) <= width(whenFlat)
+   */
+  def flatAlt(default: Doc, whenFlat: Doc): Doc =
```
Comment (johnynek) [inline core/src/main/scala/org/typelevel/paiges/Doc.scala]: I don't like making this public. So far, our API has no unsafe calls. This changes it since users must maintain the invariant.

File: core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala (hunk 1 above)
Comment (johnynek) [reply core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala]: I don't think so before. I think before `char('\n') == FlatAlt(Line, space)` I think that was defined, I don't think we should change that since it could break people's assumptions.

File: core/src/test/scala/org/typelevel/paiges/PaigesTest.scala [hunk 7]
```diff
@@ -254,4 +254,36 @@ the spaces""")
...
+    assert(origStr1 == origStr)
+
+    val origStrWide =
+      """y match { case foo => 5
+        |          case bar => 6 }""".stripMargin
+
+    val origStrWide1 = orig.render(100)
+    assert(origStrWide1 == origStrWide)
+
+    val flatStr =
```
Comment (johnynek) [inline core/src/test/scala/org/typelevel/paiges/PaigesTest.scala]: isn't this the same as `origStrWide`?

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 8]
```diff
@@ -619,7 +613,7 @@ object Doc {
    *
    * when we call `.grouped` on lineBreak we get lineOrEmpty
    */
-  val lineBreak: Doc = Line(false)
+  val lineBreak: Doc = FlatAlt(Line, Empty)
```
Comment (johnynek) [inline core/src/main/scala/org/typelevel/paiges/Doc.scala]: you had an invariant that `FlatAlt(a, b)` has `isEmpty(a) == isEmpty(b)` above, but this is false here, right?

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 9]
```diff
@@ -442,60 +444,39 @@ sealed abstract class Doc extends Product with Serializable {
...
+  private[paiges] def tryFlatten: Doc = flatten getOrElse this
 
-  /**
-   * Convert this Doc to a single-line representation.
-   *
-   * All newlines are replaced with spaces (and optional indentation
-   * is ignored). The resulting Doc will never render any newlines, no
-   * matter what width is used.
-   */
-  def flatten: Doc = flattenOption getOrElse this
```
Comment (johnynek) [inline core/src/main/scala/org/typelevel/paiges/Doc.scala]: I am concerned about removing this entirely from the API. It can't have the contract as before: all newlines removed unless we return an Option here.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala (hunk 6 above)
Comment (seanmcl) [reply core/src/main/scala/org/typelevel/paiges/Doc.scala]: Yeah, I didn't know whether to do so or not.  I was thinking that it allowed more fine grained control over custom rendering, but indeed, the docs can be totally unrelated.  

Currently users can add Text blocks with newlines and spaces, which is technically unsafe.  I agree that this is more extreme.  I am happy to make it private and add it back if it's ever needed in Scalafmt.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala (hunk 5 above)
Comment (seanmcl) [reply core/src/main/scala/org/typelevel/paiges/Doc.scala]: Ug, yes sorry.  I wanted to drop the rhs when testing for empty.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala (hunk 2 above)
Comment (seanmcl) [reply core/src/main/scala/org/typelevel/paiges/Doc.scala]: Oh, thanks for pointing that out.  I didn't appreciate it until now.  Will consider how to get it back.

**Comment by seanmcl** [comment]
Will stop updating this PR but leave this open for discussion.  I will make smaller PRs with the comments addressed.

File: core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala (hunk 1 above)
Comment (seanmcl) [reply core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala]: Oh darn it, I was conflating `text` and `Text`.  Removing.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala (hunk 6 above)
Comment (seanmcl) [reply core/src/main/scala/org/typelevel/paiges/Doc.scala]: My last comment stemmed from a basic misunderstanding of the API.  Will drop the `flatAlt` function.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala (hunk 3 above)
Comment (seanmcl) [reply core/src/main/scala/org/typelevel/paiges/Doc.scala]: To clarify, you'd like this change without the new `Line` constructor?  Then the `No` branch is unused, so I didn't think of this as totally orthogonal.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala (hunk 5 above)
Comment (seanmcl) [reply core/src/main/scala/org/typelevel/paiges/Doc.scala]: `isEmpty` seems to be only used to check the separator is empty in, e.g. `intercalate`.  In my opinion this should not be exposed in the API.  The reason is that I think of the library as being used as follows:

- Put combinators together
- Pick a column width
- Run render

I don't expect the user to be branching on whether a `Doc` is empty, or if a `Doc` can be flattened or not.  This seems to be breaking an abstraction barrier in my way of thinking about the problem.  Is this not other's view of the usage purpose of the library?  If users are doing this branching, perhaps exposing `flatAlt` would allow them to get the behavior they want without needing to hack on the `Doc`s in a non-declarative fashion.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala (hunk 8 above)
Comment (seanmcl) [reply core/src/main/scala/org/typelevel/paiges/Doc.scala]: Fixed.  Thanks for the catch.

File: core/src/test/scala/org/typelevel/paiges/PaigesTest.scala (hunk 7 above)
Comment (seanmcl) [reply core/src/test/scala/org/typelevel/paiges/PaigesTest.scala]: I cleaned this up.  Thanks for pointing out the sloppiness.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 10]
```diff
@@ -443,61 +444,40 @@ sealed abstract class Doc extends Product with Serializable {
...
+          case Nil => finish(h, front)
+          case x :: xs => loop(x, xs, h :: front)
+        }
+        case Line => No
+        case Nest(_, d) => loop(Flat(d, h.changed), stack, front) // no Line, so Nest is irrelevant
+        case Align(d) => loop(Flat(d, h.changed), stack, front) // no Line, so Align is irrelevant
+        case Union(a, _) => loop(Flat(a, true), stack, front) // invariant: flatten(union(a, b)) == flatten(a)
+        case Concat(a, b) => loop(Flat(a, h.changed), Flat(b, h.changed) :: stack, front)
+        case FlatAlt(_, b) => loop(Flat(b, true), stack, front)
+        case LazyDoc(a) => loop(Flat(a(), false), stack, front)
```
Comment (johnynek) [inline core/src/main/scala/org/typelevel/paiges/Doc.scala]: we should use `l@LazyDoc => l.evaluated` here so we cache the evaluation if it has already happened. We don't want to evaluate those more than once.

File: core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala [hunk 11]
```diff
@@ -189,26 +189,21 @@ class PaigesScalacheckTest extends FunSuite {
   }
   test("a.flatten == a.flatten.flatten") {
     forAll { (a: Doc) =>
-      val aflat = a.flatten
-      assert(aflat === aflat.flatten)
-    }
-  }
-  test("a.flatten == a.flattenOption.getOrElse(a)") {
```
Comment (johnynek) [inline core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala]: this law would be nice to add back with the tri-state FlattenResult I suggest above.

basically: `a.flatten.toDoc === a.flatten.toDoc.flatten.toDoc` which is what you have, but also:
`a.flatten.toDoc === a.flatten.orElse(a)` and also `val fa = a.flatten; if (fa.wasFlattened) fa.toDoc =!= a else fa.toDoc === a`

So, I think we want to exercise all three results of a flattening.

**Comment by johnynek** [comment]
looks like we have a law failure:
```
[info] - the left and right side of a union are the same after flattening *** FAILED *** (38 milliseconds)
[info]   TestFailedException was thrown during property evaluation.
[info]     Message: okay(a) was false
[info]     Location: (PaigesScalacheckTest.scala:209)
[info]     Occurred when passed generated values (
[info]       arg0 = Doc(...)
[info]     )
```

**Comment by johnynek** [comment]
might help to turn up minSuccessful:

https://github.com/typelevel/paiges/blob/master/core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala#L13

to maybe 5000 or something while we are refactoring and make sure we cover all the bases.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 12]
```diff
@@ -584,7 +569,7 @@ object Doc {
    *
    * Additionally, the left side (a) MUST be right associated with
    * any Concat nodes to maintain efficiency in rendering. This
-   * is currently done by flatten/flattenOption
+   * is currently done by `.grouped`.
```
[the remark this answers is not in the document]
Comment (seanmcl) [reply core/src/main/scala/org/typelevel/paiges/Doc.scala]: I'll revert.  In my mind we were dropping `flatten` from the public API so I didn't want to refer to it in the public docs.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 13]
```diff
@@ -443,61 +444,40 @@ sealed abstract class Doc extends Product with Serializable {
...
-  def flatten: Doc = flattenBoolean._1
-
-  // return the flattened doc, and if it is different
-  private def flattenBoolean: (Doc, Boolean) = {
-
-    type DB = (Doc, Boolean)
+  // Try to put the document on a single line.  Will succeed iff there are no
+  // hard line breaks.
+  // Invariant: If flatten(d) == Some(d') then d' has no FlatAlt nodes
+  private[paiges] def flatten: Option[Doc] = {
```
[the remark this answers is not in the document]
Comment (seanmcl) [reply core/src/main/scala/org/typelevel/paiges/Doc.scala]: That seems good.  I have a couple hesitations

- We seem to be depending on the fact that `Concat`s are right associated for performance, but that is only handled here, and it is not certain the caller has called `grouped` or `flatten` before `render`.  I was thinking perhaps there should be a `normalize` method that does that on all documents, and that `render` either calls `normalize` or ensures the `Doc` was normalized, e.g. with a phantom type.  

- It seems weird to throw all flattening away if there is a single hardline in the document.  I tried a couple approaches to flattening while leaving the hardline in, but both failed to handle the simple Scala formatting examples.  

So I'm still thinking a bit how to do this better, but your solution works, and if we think of something better in the future we can always do it later.

**Comment by seanmcl** [comment]
I added `Fail` in addition to `FlatAlt`, which simplifies the change considerably.  There's some remaining wonkiness, e.g. not knowing what to do with a `Fail` in the  `isEmpty` method, but I think this is a better approach overall.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 14]
```diff
@@ -514,12 +520,13 @@ sealed abstract class Doc extends Product with Serializable {
     @tailrec
     def loop(pos: Int, lst: List[(Int, Doc)], max: Int): Int = lst match {
       case Nil => math.max(max, pos)
-      case (i, Empty) :: z => loop(pos, z, max)
+      case (i, Empty | Fail) :: z => loop(pos, z, max)
```
Comment (johnynek) [inline core/src/main/scala/org/typelevel/paiges/Doc.scala]: seems like we should not just silently skip `Fail` to me. Can you explain in a comment why we should?

Seems to me that we want `Union(Fail, a) == a` and otherwise we don't expect to see `Fail`. Actually, I would say `Fail` should actually throw in rendering since by invariant we expect it to never come up.

File: core/src/main/scala/org/typelevel/paiges/Doc.scala [hunk 15]
```diff
@@ -537,13 +544,25 @@ object Doc {
...
+   * Invariant: default != whenFlat
+   */
+  private[paiges] case class FlatAlt(default: Doc, whenFlat: Doc) extends Doc
+
+  /**
+   * Used during rendering to indicate a hard line break which
+   * can't be flattened.
+   * Invariant: Fail only appears on the left-hand side (i.e. the flattened side) of a Union
+   */
+  private[paiges] case object Fail extends Doc
```
Comment (johnynek) [inline core/src/main/scala/org/typelevel/paiges/Doc.scala]: can't this be returned by `def flatten`? We should have a test for this invariant:

```scala
forAll { a: Doc =>
  // something like this except a recursive check:
  a match {
    case Union(Fail, _) => assert(true)
    case Fail => assert(false)
    case _ => assert(true)
  }

  // more over:
  assert(a.flatten != Fail)
}
```

Can you describe how this can be rendered? And what it maxWidth, etc... is?

We should have stronger tests if we are adding a new variant.

Comment (seanmcl) [reply core/src/main/scala/org/typelevel/paiges/Doc.scala]: Yes, `Fail` is frequently returned by `flatten`, as it is the instruction to `best` to ignore the current path, and use the next one.   

I'm trying to figure out what the laws should be now.  We don't need to add `Fail` randomly to the generators, since it's only inserted by `Flatten`.  `FlatAlt` currently only comes from `line`, etc, so I'll throw those in.  But indeed, we need to clarify in the laws where `Fail` is allowed.  I'll submit another patch shortly.

**Comment by seanmcl** [comment]
Dropping this.  Will publish a clean solution in another PR taking these comments into account.
